include/coreir/ir/wireable.h

```diff
@@ -128,12 +129,15 @@ class Select : public Wireable {
   protected :
     Wireable* parent;
     std::string selStr;
+    bool _isInstance;
```

Comment:
Is this trying to differentiate between a hierarchical select and a normal select? I think the hierarchical select would be better as a separate class (inherited from Select). I am a little worried that this might break code which recursively goes through all the selects of an instance (or interface).

Reply:
What would break? This provides the same interface for existing selects, so it shouldn’t break any existing code, it just adds another interface function.

Reply:
i can see how much code it would take to change to add a new select class, my main concern is there’s a lot of code that uses Select so adding a new class to the code path will require a lot of changes to dispatch on the two types of selects, so we should decide whether it’s worth treating them separately versus treating a select as a single entity with a unified interface.

Reply:
i can see how much code it would take to change to add a new select class, my main concern is there’s a lot of code that uses Select so adding a new class to the code path will require a lot of changes to dispatch on the two types of selects, so we should decide whether it’s worth treating them separately versus treating a select as a single entity with a unified interface.

Reply:
The thing I am concerned about is the following:

Lets say I have Module A instantiating Module B which instantiates Module C
Additionally lets say within A, I have a connection referring to the instance of C within A's instance of B.

During an optimization pass of B I do not want to have access to the select of C and the fact that some unknown Module (A) is using that select.

Reply:
That could be achieved by filtering the selects (in the function to gather connected wires led) rather than encoding it in the class system. There might be cases when you do want to know about the connection (e.g. if you’re removing the port you’ll probably want to remove that connection in another context?) that brings up a question about how a hierarchical selects will affect optimization passes

Reply:
Another option is to have magma lift these hierarchical selects into the interface and drop support for it in coreir directly, which may make sense given the above issue with optimization passes (cc @rsetaluri, any thoughts? We originally discussed this as something that might be useful in coreir, but it may introduce semantic complexity we dint want to have in the IR)

Reply:
I think this feature is a good thing to have in CoreIR and by lifting it into the interface directly, you are losing information that could be passed down and are modifying interfaces unnecessarily.

I am just trying to enumerate possible issues with this.

Reply:
Okay I think the filtering would work, but there needs to be an update to the logic so that any Select from a hierarchical select is also a hierarchical select. Also making sure that the connection does not appear in the Module with the hierarchical select.

Reply:
Another thought, do we want to enable connecting to inputs hierarchically? For example, one use case is to monitor an internal signal, so we can reference an output and send it into a debug/verification node.  Do we want to support users wiring to inputs hierarchically? This may conflict with our current definition semantics, where everything must be driven or termed.  essentially, you would leave an input undriven, and it could be wired up from the instance location.  In this case, we'd want to run the undriven check at the end on the entire hierarchy rather than per module (not sure if this is what is already done).  Otherwise, we should update the connect logic to make sure hierarchical selects only reference outputs.  I believe the current use case only wants to monitor signals, so this may be a good simplification to start with.

Reply:
Yeah, I think we should constrain hierarchical selects to outputs for now (as that is the pressing use case). Driving hierarchical selects opens up a larger can of worms and we should do a review of tools that allow this.

Reply:
I confirmed that the users only want to monitor signals, and that they aren't aware of any patterns for driving signals this way, so I don't think we need to support it (at least until a valid use case comes up).

Perhaps for now we can treat these a special selects for debug nodes? That might help avoid having them interfere with standard "RTL" code/nodes

Reply:
That could work. How about for now to simplify this just instantiate debug nodes, do not connect them to anything, and store the hierarchical select path as the debug node's metadata or modargs.

Reply:
yea that seems like a good idea and less invasive for current code paths.  I'll resubmit a new PR with this change as it will require a different overall approach.

Reply:
Actually, there's still the issue of maintaining the select path when optimizations are run (e.g. what if a monitored signal gets renamed in an optimization, ideally we'd still be able to connect to that signal).  That's why I wanted to piggy back on the connection logic, since ideally that'd be kept up to date

Reply:
So in some sense, optimizations should be aware of these "external" hierarchical selects, if just to make sure that their connections are preserved

Reply:
Yeah, I understand the concern.

I think the biggest constraint is that CoreIR assumes that Modules and module Defs are completely independent of the outside world (namely other modules that could instantiate them). Therefore connections should always be between two wireables within the same parent Module.

Here is a half baked idea that I think could still work with your constraints. What if we have some sort of "portal" node that can go between module boundaries without going through the interface.

So for example in the A instances B which instances C.
You can instantiate one end of the portal in A (and connect it to the debug node), then also instantiate a portal node in B and (connect it to the appropriate signal within B). Any optimizations in B (even renaming signals) should not effect the result. The only issue is that optimizations in A (like renaming the instance of B) might pose a problem.

Reply:
I was thinking about this some more, I think for debug/verification purposes, having a reference to a signal should prevent the signal from being optimized out.  So a subtract could get turned into an an add with invert, so long as the original signal reference now points to the output of the add.  I guess with common sub expression elimination, if the signal mapped to the common sub expression output that would be okay too.  Is there ever a case where a signal might be removed entirely? (Without a new mapping)

Reply:
Yeah, as long as there is something connected to the signal you care about, then any optimizations should still work. There is a chance that a particular signal might get optimized away. For example if you are trying to poll the output of an Add(a, 0). If this gets optimized away without the module knowing that something is connected to its output, then the reference is lost.

Reply:
This is why if there is something that the desired signal is connected to a portal or term node of some sorts, then the reference will be updated appropriately during any optimizations.

Reply:
I think for now, I will see if I can add a separate class for InstanceSelect otherwise stick to the boolean flag on the Select type, and use a symbol table to handle renaming by optimizations downstream, and if needed we can add a filter for instance selects when gathering connected wireables.